[AMDGPU] Remove SIWholeQuadMode pseudo wavemode optimization

perlfu opened the pull request:
This does not work correctly in divergent control flow. Can be replaced with a later exec mask manipulation optimizer.

This reverts commit a3646ec1bc662e221c2a1d182987257c50958789.

perlfu committed:
[AMDGPU] Remove SIWholeQuadMode pseudo wavemode optimization

This does not work correctly in divergent control flow.
Can be replaced with a later exec mask manipulation optimizer.

This reverts commit a3646ec1bc662e221c2a1d182987257c50958789.

piotrAMD commented:
Code changes look good as this is a clean revert.

What is "a later exec mask manipulation optimizer" ?
Would it be possible to post a simple test case too?

perlfu commented:
> Code changes look good as this is a clean revert.

It is technically one line different from a direct revert, because the original commit fixed a one line bug.

> Would it be possible to post a simple test case?

The existing if-endif cases kind of cover this, but I have added a new test which shows the issue with kill call in WQM shaders and strict WQM.

> What is "a later exec mask manipulation optimizer"?

I think there is scope for adding a exec mask analysis which could detect regions where the mask would be the same and removed unnecessary changes.
This would be separate follow up work which would safely reimplement the optimization removed with this PR.

arsenm commented:
> The existing if-endif cases kind of cover this, but I have added a new test which shows the issue with kill call in WQM shaders and strict WQM.
> 

I don't see the new test?

piotrAMD commented:
> > The existing if-endif cases kind of cover this, but I have added a new test which shows the issue with kill call in WQM shaders and strict WQM.
> 
> I don't see the new test?

Pre-committed in https://github.com/llvm/llvm-project/commit/a44d7406f45fd3e5af45de116aed03b0bf7a881f.

piotrAMD reviewed (approved):
LGTM
